proposals/0010-resource-element-type-validation.md

```diff
@@ -0,0 +1,136 @@
+* Proposal: [0010](0010-resource-element-type-validation.md)
+* Author(s): [Joshua Batista](https://github.com/bob80905)
+* Sponsor: Joshua Batista
+* Status: **Under Consideration**
+* Impacted Project(s): (LLVM)
+
+* Issues: [#75676](https://github.com/llvm/llvm-project/issues/75676)
+
+## Introduction
+Resources are often used in HLSL, with various resource element types (RETs).
+For example:
+```
+RWBuffer<float> rwbuf: register(u0);
+```
+In this code, the RET is `float`, and the resource type is `RWBuffer`. The
+resource type is not a `RawBuffer` variant, and so there is a distinct set
+of rules that define valid RETs for this resource type.
+
+RETs for non-`RawBuffer` variants may include basic types (ints and uints of sizes 16
+and 32, as well as half, float, vectors, and matrices of 4 elements or fewer).
+Structs that contain fields of these basic types (where all fields in the struct have
+the same type) may also be RETs. 
+Structs that either have structs as fields or arrays of structs as fields may also be
+allowed, as long as there are at most 4 sub elements, and each sub element is at most
+32 bits. Additionally, resource types are not allowed within an RET, even if the 
+underlying resource type has a primitive RET (i.e., `RWBuffer<int>` as an RET).

```

Comment:
I find these big blocks of prose really hard to read.

```suggestion

RETs for non-`RawBuffer` variants may include:
* basic types: 
  * 16- and 32-bit int and uint
  * half and float
  * vectors
  * matrices containing 4 elements or fewer
* user defined types (structs / classes), as long as:
  * all fields in the struct have the same type
  * there are at most 4 sub elements, and each sub element is at most 32 bits
  
Resource types are not allowed within an RET, even if the underlying resource
type has a primitive RET (i.e., `RWBuffer<int>` as an RET).
```

Reply:
I feel that there must be simpler way of explaining this...isn't it just "max of 4 elements and max of 128 bits"?

Maybe some examples could be used to illustrate how that rule applies, but I think there's just one rule?

Reply:
Used bulleted list instead.
If I were to use "max of 4 elements and max of 128 bits", a user may think 2 64-bit values would work. This would not work, it would cause undefined behavior at runtime. We must specify that each element may not exceed 32 bits.

Reply:
Bulletted list looks much better! I'm going to mark this as resolved and cover the 64-bit value support in a new conversation.